kv/util/codec/codec.go

```diff
@@ -0,0 +1,121 @@
+package codec
+
+import (
+	"encoding/binary"
+
+	"github.com/pingcap/errors"
+)
+
+const (
+	signMask uint64 = 0x8000000000000000
+
+	encGroupSize = 8
+	encMarker    = byte(0xFF)
+	encPad       = byte(0x0)
+)
+
+var pads = make([]byte, encGroupSize)
+
+// EncodeKey encodes a user key and appends an encoded timestamp to a key. Keys and timestamps are encoded so that
+// timestamped keys are sorted first by key (ascending), then by timestamp (descending). The encoding is based on
+// https://github.com/facebook/mysql-5.6/wiki/MyRocks-record-format#memcomparable-format.
+func EncodeKey(key []byte, ts uint64) []byte {
+	encodedKey := EncodeBytes(key)
+	return AppendTs(encodedKey, ts)
+}
+
+// EncodeBytes guarantees the encoded value is in ascending order for comparison,
+// encoding with the following rule:
+//  [group1][marker1]...[groupN][markerN]
+//  group is 8 bytes slice which is padding with 0.
+//  marker is `0xFF - padding 0 count`
+// For example:
+//   [] -> [0, 0, 0, 0, 0, 0, 0, 0, 247]
+//   [1, 2, 3] -> [1, 2, 3, 0, 0, 0, 0, 0, 250]
+//   [1, 2, 3, 0] -> [1, 2, 3, 0, 0, 0, 0, 0, 251]
+//   [1, 2, 3, 4, 5, 6, 7, 8] -> [1, 2, 3, 4, 5, 6, 7, 8, 255, 0, 0, 0, 0, 0, 0, 0, 0, 247]
+// Refer: https://github.com/facebook/mysql-5.6/wiki/MyRocks-record-format#memcomparable-format
+func EncodeBytes(data []byte) []byte {
+	// Allocate more space to avoid unnecessary slice growing.
+	// Assume that the byte slice size is about `(len(data) / encGroupSize + 1) * (encGroupSize + 1)` bytes,
+	// that is `(len(data) / 8 + 1) * 9` in our implement.
+	dLen := len(data)
+	result := make([]byte, 0, (dLen/encGroupSize+1)*(encGroupSize+1)+8) // make extra room for appending ts
+	for idx := 0; idx <= dLen; idx += encGroupSize {
+		remain := dLen - idx
+		padCount := 0
+		if remain >= encGroupSize {
+			result = append(result, data[idx:idx+encGroupSize]...)
+		} else {
+			padCount = encGroupSize - remain
+			result = append(result, data[idx:]...)
+			result = append(result, pads[:padCount]...)
+		}
+
+		marker := encMarker - byte(padCount)
+		result = append(result, marker)
+	}
+	return result
+}
+
+// AppendTs appends the timestamp to encoded key, Note we invert the timestamp so that when sorted, they are in descending order.
+func AppendTs(encodedKey []byte, ts uint64) []byte {
```

Comment:
Do we need this function? I think it is better to just have EncodeKey because I don't think users of this package should have an encoded key without the timestamp (also, the name is a bit misleading since we 'encode' and append the TS, not just append).